Dapr state store clickhouse

middt opened the pull request:
# Description

This PR adds a new state store component for ClickHouse, a column-oriented database management system. The ClickHouse state store component provides the following features:

- Basic CRUD operations for state management
- Support for TTL (Time-To-Live)
- ETag support for optimistic concurrency
- Bulk operations support
- Username/password authentication
- Uses ReplacingMergeTree engine for efficient updates

Key implementation details:
- Uses the official ClickHouse Go driver
- Implements the full state.Store interface
- Handles connection management and cleanup
- Includes comprehensive integration tests
- Follows Dapr's component development guidelines

## Implementation Details

The component includes:
1. State store implementation (`clickhouse.go`)
   - Full CRUD operations
   - Bulk operations
   - TTL support
   - ETag support
   - Authentication support

2. Tests (`clickhouse_test.go`)
   - Integration tests for CRUD operations
   - Metadata validation tests
   - Authentication tests
   - Cleanup handling

3. Configuration options:
   ```yaml
   apiVersion: dapr.io/v1alpha1
   kind: Component
   metadata:
     name: statestore
     namespace: default
   spec:
     type: state.clickhouse
     version: v1
     metadata:
     - name: clickhouseURL
       value: "tcp://localhost:9000"
     - name: databaseName
       value: "dapr_state"
     - name: tableName
       value: "state_store"
     - name: username
       value: "default"
     - name: password
       value: ""
   ```

## Checklist

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

## Testing Done

1. Unit tests for metadata validation
2. Integration tests for:
   - Basic CRUD operations
   - TTL functionality
   - ETag support
   - Bulk operations
   - Authentication
   - Error handling

Tests were run against ClickHouse v23.8 using the official Go driver.

## Additional Notes

- The component uses ReplacingMergeTree engine for better update performance
- Includes proper cleanup in tests to avoid test pollution
- Follows ClickHouse best practices for table design
- Implements proper connection management and cleanup

JoshVanL commented:
Thanks @middt, implementation generally looks good to me!

Would it be possible to add [conformance](https://github.com/dapr/components-contrib/blob/main/tests/conformance/state_test.go) tests for this too? It seems like we should be able to spin up a [click house docker compose](https://github.com/dapr/components-contrib/tree/main/.github/infrastructure).

middt commented:
Thank you for the feedback! @JoshVanL  I've already implemented the conformance tests for the ClickHouse state store in this PR:

1. Created a Docker Compose setup in `/tests/certification/state/clickhouse/docker-compose.yml` that spins up a ClickHouse server with proper configuration
2. Added a conformance test file in `/tests/certification/state/clickhouse/clickhouse_test.go` that verifies all state store operations
3. Created configuration files in `/tests/config/state/clickhouse/clickhouse.yml` for the test environment
4. Added ClickHouse to the state store tests configuration in `/tests/config/state/tests.yml`

The tests verify all the key functionality including:
- Basic CRUD operations
- ETag support for optimistic concurrency
- Error handling for empty and non-existent keys
- Feature reporting

I've also addressed authentication issues by properly configuring username and password in both the Docker Compose setup and the state store implementation.

All unit tests are now passing, confirming that the implementation works correctly with the ClickHouse server.

middt commented:
Done! Just ran `go mod tidy` and pushed the changes. Thanks for the review and guidance! 👍 

@JoshVanL

cicoyle commented:
@middt - mind resolving the conflicts with `go.mod` & `go.sum`?

middt commented:
> Also, mind checking the linter locally? It appears to fail in CI



> Also, mind checking the linter locally? It appears to fail in CI

Thanks for the feedback! @cicoyle   I've already fixed all linter issues locally (import formatting, gosec warnings, variable shadowing, etc.) and the latest commits should pass CI.

middt commented:
ping pong

middt commented:
> Hey @middt, tests are failing

Hi @JoshVanL,
Fixed! Tests now pass:, Ready for review!

sicoyle commented on state/clickhouse/clickhouse.go:
```diff
@@ -0,0 +1,435 @@
+		return config, errors.New("ClickHouse database name is missing")
+	}
+
+	if config.TableName == "" {
```
are there fixed char limits on these we should check for by chance?

CasperGN replied:
@sicoyle [looks like it's here](https://clickhouse.com/docs/sql-reference/functions/other-functions#getMaxTableNameLengthForDatabase).
@middt can you add a length check as well?

sicoyle commented on state/clickhouse/metadata.yaml:
```diff
@@ -0,0 +1,50 @@
+        example: "mypassword"
+        default: ""
+metadata:
+  - name: ClickhouseURL
```
can you pls update these all to be camelcase?

sicoyle commented on tests/certification/state/clickhouse/clickhouse_test.go:
```diff
@@ -0,0 +1,155 @@
+// +build conftests
+
+/*
+Copyright 2021 The Dapr Authors
```
```suggestion
Copyright 2025 The Dapr Authors
```

sicoyle reviewed (commented):
few comments. Nice job 👏

CasperGN commented:
@sicoyle I think this one is actually super interesting to get in. I've seen an uptick in Clickhouse lately so supporting this would be beneficial imo.
@middt please address the comments so we can get this one merged 😄

CasperGN commented on state/clickhouse/clickhouse.go:
```diff
@@ -39,12 +39,15 @@ type StateStore struct {
 }
 
+// maxIdentifierLength is the maximum length for database and table names in ClickHouse
+const maxIdentifierLength = 256
```
[Per their docs](https://clickhouse.com/docs/sql-reference/functions/other-functions#getMaxTableNameLengthForDatabase) and [this issue](https://github.com/ClickHouse/ClickHouse/issues/64861) they reference a length of `206` and `250` respectively - are you positive about the `256` limit?

CasperGN replied:
paging @middt

CasperGN commented on .github/workflows/dependabot-tidy.yml:
```diff
@@ -28,6 +26,7 @@ jobs:
               run: |
                 git config --global user.name "github-actions[bot]"
                 git config --global user.email "github-actions[bot]@users.noreply.github.com"
+                git remote set-url origin https://x-access-token:${{ secrets.GITHUB_TOKEN }}@github.com/$GITHUB_REPOSITORY
```
This is also removed on main

CasperGN reviewed (changes requested):
@middt it looks like the merge from main reverted a bunch of stuff. Can you please look over it again?

middt committed:
feat(state): add ClickHouse state store component

Add a new state store component for ClickHouse with:
- Full CRUD operations via database/sql with clickhouse-go driver
- ETag support for optimistic concurrency (UUID-based)
- TTL support with DateTime64 expiration column
- ReplacingMergeTree engine for efficient updates
- Username/password authentication
- Bulk operations (BulkGet, BulkSet, BulkDelete)
- Identifier length validation (max 250 for filesystem safety)
- Auto-creation of database and table if not exists
- Conformance test configuration and Docker Compose infrastructure
- Certification test with Docker Compose setup

Addresses review feedback:
- Copyright year set to 2025
- Metadata fields use camelCase (clickhouseUrl, databaseName, tableName)
- maxIdentifierLength reduced to 250 (from 256) for filesystem suffix room
- Clean rebase on main (no accidental reverts of unrelated files)

Signed-off-by: Mehmet TOSUN <93265833+middt@users.noreply.github.com>

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse

dapr-bot committed:
Merge branch 'main' into dapr-state-store-clickhouse
